uc-crux-llvm/src/UCCrux/LLVM/Postcondition/Type.hs

```diff
@@ -0,0 +1,275 @@
+    -- | Specification of the return value
+    _uReturnValue :: Maybe (ConstrainedTypedValue m)
+  }
+
+uArgClobbers :: Lens' (UPostcond m) (IntMap (SomeClobberArg m))
+uArgClobbers = Lens.lens _uArgClobbers (\s v -> s { _uArgClobbers = v })
+
+uGlobalClobbers :: Lens' (UPostcond m) (Map (GlobalSymbol m) (SomeClobberValue m))
+uGlobalClobbers = Lens.lens _uGlobalClobbers (\s v -> s { _uGlobalClobbers = v })
+
+uReturnValue :: Lens' (UPostcond m) (Maybe (ConstrainedTypedValue m))
+uReturnValue = Lens.lens _uReturnValue (\s v -> s { _uReturnValue = v })
+
+-- | A postcondition which imposes no constraints and has a no return value (is
+-- void).
+emptyUPostcond :: UPostcond m
+emptyUPostcond =
+  UPostcond
+    { _uArgClobbers = IntMap.empty,
+      _uGlobalClobbers = Map.empty,
+      _uReturnValue = Nothing
+    }
+
+-- | A postcondition which imposes no constraints and has some
+-- minimally-constrained return value.
+minimalUPostcond :: ReturnTypeRepr m ft -> UPostcond m
+minimalUPostcond retRepr =
+  UPostcond
+    { _uArgClobbers = IntMap.empty,
+      _uGlobalClobbers = Map.empty,
+      _uReturnValue =
+        case retRepr of
+          FS.VoidRepr -> Nothing
+          FS.NonVoidRepr ft ->
+            Just (ConstrainedTypedValue ft (minimalConstrainedShape ft))
+    }
+
+ppUPostcond :: UPostcond m -> PP.Doc Void
+ppUPostcond post =
+  PP.vsep $ bullets
+    [ "Return value:" PP.<+>
+        case _uReturnValue post of
+          Just (ConstrainedTypedValue _type shape) -> ppConstrainedShape shape
+          Nothing -> "<void>"
+    , header
+        "Argument clobbers:"
+        (map (uncurry ppArg) (IntMap.toList (_uArgClobbers post)))
+    , header
+        "Global clobbers:"
+        (map (uncurry ppGlob) (Map.toList (_uGlobalClobbers post)))
+    ]
+  where
+    bullets = map ("-" PP.<+>)
+    header hd items = PP.nest 2 (PP.vsep (hd : bullets items))
+
+    ppArg i (SomeClobberArg ca) =
+      (PP.viaShow i <> ":") PP.<+>
+        case ca of
+          DontClobberArg -> "no clobbering"
+          DoClobberArg cv -> ppClobberValue cv
+
+    ppGlob gSymb (SomeClobberValue cv) =
+      (PP.viaShow (getGlobalSymbol gSymb) <> ":") PP.<+> ppClobberValue cv
+
+data ReturnValue m mft f where
+  ReturnVoid :: ReturnValue m 'Nothing f
+  ReturnValue :: f ft -> ReturnValue m ('Just ft) f
+
+-- | A more strongly typed version of 'UPostcond'.
+data Postcond m (fs :: FuncSig m) where
+  Postcond ::
+    { pVarArgs :: VarArgsRepr va,
+      -- | Specifications of which pointer arguments are written to, and how
+      pArgClobbers :: Assignment (ClobberArg m) argTypes,
+      -- | Specifications of which global variables are written to, and how
+      pGlobalClobbers :: Map (GlobalSymbol m) (SomeClobberValue m),
+      -- | Specification of the return value
+      pReturnValue :: ReturnValue m mft (ConstrainedShape m)
+    } ->
+    Postcond m ('FS.FuncSig va mft argTypes)
+
+data PostcondTypeError
```

Comment:
Can we capture more context here to help interpret these errors? For example, the two types that don't match? Is there enough context at the error handling site to note the function that generated the error?

Reply:
I've added more detail. Also, there's only one callsite at the moment, which panics and prints out the function signature and postcondition which is always sufficient to see what the problem was (perhaps not immediately/easily - but then again, it's only shown in a panic).